### PR TITLE
uiWindow unit tests and fixes for unix

### PR DESCRIPTION
**Pull request by szanni**
This sadly got a lot bigger than initially anticipated but much of this code is interdependent, hence all of this at once: 

- Add unit tests for uiWindow
- Document uiWindowMargined default to be `FALSE`

Fixes for unix:
- Fix uiWindowContentSize upon window creation, see #252
- Make uiWindowSetPosition non blocking when hidden
- Make uiWindowSetContentSize non blocking when hidden
- Fix uiWindowSetContentSize for margined windows

`uiWindowContentSize` now uses  `gtk_window_get_size` and `gtk_window_resize` instead of trying to do some magic calculations on the inner container. According to the docs gtk already does a best effort calculation to report the size sans window decorations. The prior, inner container size code does not work on hidden windows and margined windows anyways and does not add anything on all systems I tested (i3wm and weston) apart from code, confusion, bugs and brittleness, hence the removal.

It is unclear to me if there was ever a good reason for doing these calculations. I could not find any documentation in the code or git log. All code related to window sizing is a best effort on all Unix systems anyways as the [gtk3 docs](https://docs.gtk.org/gtk3/method.Window.get_size.html) state everywhere. I did not notice any regressions by removing all the window size calculations.

### CODE REVIEW

File: unix/main.c [hunk 1]
```diff
@@ -19,6 +19,10 @@ const char *uiInit(uiInitOptions *o)
 	uiprivInitAlloc();
 	uiprivLoadFutures();
 	timers = g_hash_table_new(g_direct_hash, g_direct_equal);
+
+	// Run the event loop manually by default to ensure we can run uiMainStep()
+	// internally to make asynchronous GTK calls appear synchronous.
+	uiMainSteps();
```
Comment (cody271) [inline unix/main.c:25]: Shouldn't the event loop default to "steps mode" on Darwin now as well?

**Review by cody271 — COMMENTED** [review]
Has this been tested on Wayland yet?

**Comment by cody271** [comment]
> It is unclear to me if there was ever a good reason for doing these calculations. I could not find any documentation in the code or git log. All code related to window sizing is a best effort on all Unix systems anyways as the [gtk3 docs](https://docs.gtk.org/gtk3/method.Window.get_size.html) state everywhere. I did not notice any regressions by removing all the window size calculations.
>
👍 

>> on all systems I tested (i3wm and weston)
>>
> Has this been tested on Wayland yet?

I will try some other DEs besides Weston.

File: unix/main.c (hunk 1 above)
Comment (szanni) [reply unix/main.c:25]: > Shouldn't the event loop default to "steps mode" on Darwin now as well?

I guess that would make sense for consistencies sake? On the other hand Windows has been doing it's own thing all along, which has been calling `uiMainStep()` in `uiMain()`.

Happy to change that, just needs to be tested very thoroughly to make sure we don't break anything. There seems to be some interaction around that and uiMenu.

**Comment by szanni** [comment]
> > > on all systems I tested (i3wm and weston)
> > 
> > Has this been tested on Wayland yet?
> 
> I will try some other DEs besides Weston.

To add to that, this means the unit test work. By no means do Wayland systems honor `uiWindowSetPosition` in any shape or form as Wayland [has no protocol for doing so](https://gitlab.freedesktop.org/wayland/wayland-protocols/-/merge_requests/247). And SetContentSize only really works when increasing window size form the original on Wayland systems consistently.
Tested with weston and sway+floating windows.

**Comment by kojix2** [comment]
Hi szanni.
Thank you for creating this pull request. Even before merging this pull request into libui-ng/master, our Ruby team can distribute the pre-package that includes it.

In this way, if any new problems occur, there is a good chance that someone will find and report them.
If you don't mind, we would be grateful if you could resolve the conflicts. Thank you.

**Comment by cody271** [comment]
> Tested with weston and sway+floating windows.
>

Confirmed tests passing on both GNOME/Wayland and Weston:

```
[==========] Running 11 test(s).
[ RUN      ] windowNew
[       OK ] windowNew
[ RUN      ] (windowDefaultMargined)
[       OK ] (windowDefaultMargined)
[ RUN      ] (windowMargined)
[       OK ] (windowMargined)
[ RUN      ] (windowDefaultTitle)
[       OK ] (windowDefaultTitle)
[ RUN      ] (windowDefaultContentSize)
[       OK ] (windowDefaultContentSize)
[ RUN      ] (windowSetTitle)
[       OK ] (windowSetTitle)
[ RUN      ] (windowSetPosition)
[       OK ] (windowSetPosition)
[ RUN      ] (windowSetContentSize)
[       OK ] (windowSetContentSize)
[ RUN      ] (windowMarginedSetContentSize)
[       OK ] (windowMarginedSetContentSize)
[ RUN      ] (windowSetContentSizeNoCallback)
[       OK ] (windowSetContentSizeNoCallback)
[ RUN      ] (windowSetPositionNoCallback)
[       OK ] (windowSetPositionNoCallback)
[==========] 11 test(s) run.
[  PASSED  ] 11 test(s).
```

**Review by cody271 — APPROVED** [review]
LGTM

**Pull request merged**